### PR TITLE
#22 - newsletter

### LINKED ISSUE #22: Newsletter
For now it can be just simple sending emails. Later we can extend it, i.e. sending personalized emails or so.

### PR DESCRIPTION
**Pull request by Majkulnn**
I have created a basic newsletter page that store newsletter subscribers in database

- model and migration;
- controller and form request;
- basic routing;
- basic views;

Should close #22

### CODE REVIEW

File: database/migrations/2022_03_29_120757_create_newsletter_subscribers_table.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+declare(strict_types=1);
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration {
+    /**
+     * Run the migrations.
+     */
```
Comment (kamilpiech97) [inline database/migrations/2022_03_29_120757_create_newsletter_subscribers_table.php:12]:
```suggestion
```

Comment (kamilpiech97) [reply database/migrations/2022_03_29_120757_create_newsletter_subscribers_table.php:12]: Remove that comments.

File: database/migrations/2022_03_31_162450_create_newsletter_preferences_table.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                ->references("id")->on("newsletter_subscribers")
+                ->onDelete("cascade");
+            $table->string("preference");
+            $table->timestamps();
+        });
+    }
+
+    /**
+     * Reverse the migrations.
+     */
```
Comment (kamilpiech97) [inline database/migrations/2022_03_31_162450_create_newsletter_preferences_table.php:27]: The same as above

File: resources/views/newsletter/index.blade.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                form.submit();
+            }
+            function unsubscribe() {
+                form.action="{{ route('newsletter.destroy') }}";
+                form.submit();
+            }
+
+        </script>
+    </div>
+@endsection
```
Comment (kamilpiech97) [inline resources/views/newsletter/index.blade.php:34]:
```suggestion
@endsection

```

File: resources/views/newsletter/subscribe.blade.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+                <label for="News">New News
+                    <input type="checkbox" name="type[]" id="News" checked="checked" value={{\Blumilk\Meetup\Core\Enums\AvailableNewsletter::NEWS->value}}></label>
+                <label for="News">New Meetups
+                    <input type="checkbox" name="type[]" id="Meetups" value="{{\Blumilk\Meetup\Core\Enums\AvailableNewsletter::MEETUPS->value}}"></label>
+                <x-input-error for="type"/>
+            </div>
+            <input type="submit" name="Subscribe" value="Subscribe">
+        </form>
+    </div>
+@endsection
```
Comment (kamilpiech97) [inline resources/views/newsletter/subscribe.blade.php:25]:
```suggestion
@endsection

```

File: resources/views/newsletter/subscribe.blade.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        @endif
+        <form action="{{route("newsletter.update")}}" method="post">
+            @csrf
+            <input type="email" hidden="hidden"  name="email" id="email" value="{{ $subscriber->email }}">
+            <label for="Notification type">Notification type:</label>
+            <div>
+                <label for="News">New News
+                    <input type="checkbox" name="type[]" id="News" checked="checked" value={{\Blumilk\Meetup\Core\Enums\AvailableNewsletter::NEWS->value}}></label>
+                <label for="News">New Meetups
+                    <input type="checkbox" name="type[]" id="Meetups" value="{{\Blumilk\Meetup\Core\Enums\AvailableNewsletter::MEETUPS->value}}"></label>
```
Comment (kamilpiech97) [inline resources/views/newsletter/subscribe.blade.php:19]: Add alias for that Enum in `config/app.php`.

File: src/Http/Requests/Newsletter/NewsletterUpdateRequest.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class NewsletterUpdateRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "email" => ["required"],
+            "type" => ["required"],
```
Comment (kamilpiech97) [inline src/Http/Requests/Newsletter/NewsletterUpdateRequest.php:15]: Maybe we should validate the type more precisely? What I see, we have in Enum two types.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: database/migrations/2022_03_29_120757_create_newsletter_subscribers_table.php (hunk 1 above)
Comment (krzysztofrewak) [reply database/migrations/2022_03_29_120757_create_newsletter_subscribers_table.php:12]: @kamilpiech97, maybe you could implement something like this for our Codestyle package? https://github.com/kubawerlos/php-cs-fixer-custom-fixers/blob/main/src/Fixer/NoDoctrineMigrationsGeneratedCommentFixer.php

File: src/Http/Requests/Newsletter/NewsletterStoreRequest.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+namespace Blumilk\Meetup\Core\Http\Requests\Newsletter;
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class NewsletterStoreRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "email" => ["required"],
```
Comment (kamilpiech97) [inline src/Http/Requests/Newsletter/NewsletterStoreRequest.php:14]: Only `required` rule, maybe some validation of email structure.

File: src/Http/Requests/Newsletter/NewsletterUpdateRequest.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+use Blumilk\Meetup\Core\Enums\AvailableNewsletter;
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Validation\Rule;
+
+class NewsletterUpdateRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "email" => ["required"],
```
Comment (kamilpiech97) [inline src/Http/Requests/Newsletter/NewsletterUpdateRequest.php:16]: Same as above.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Services/NewsletterService.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+use Illuminate\Http\Response;
+
+class NewsletterService
+{
+    public function isSubscriber(NewsletterSubscriber $subscriber): Response
+    {
+        if ($subscriber->getAttribute("status") === 1) {
+            return response("You are already subscriber. Would you like to change your preferences?");
+        }
+        $subscriber->status = 1;
```
Comment (kamilpiech97) [inline src/Services/NewsletterService.php:17]: Why are you using a number here in the status field, and below is a boolean value (`unsubscribe` method - `false`)

**Review by kamilpiech97 — APPROVED** [review]

File: src/Models/News.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Models;
+
+use Blumilk\Meetup\Core\Formats;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+
+class News extends Model
```
Comment (MrHDOLEK) [inline src/Models/News.php:11]: I think a better model name is a Newsletter

File: src/Models/News.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        "content",
+    ];
+    protected $casts = [
+        "date:" . Formats::DATETIME,
+    ];
+
+    public function user(): BelongsTo
+    {
+        return $this->belongsTo(User::class);
+    }
```
Comment (MrHDOLEK) [inline src/Models/News.php:26]: there is no reference to factory

File: src/Services/NewsletterService.php [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+namespace Blumilk\Meetup\Core\Services;
+
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+use Illuminate\Http\Response;
+
+class NewsletterService
+{
+    public function isSubscriber(NewsletterSubscriber $subscriber): Response
+    {
+        if ($subscriber->getAttribute("status") === 1) {
```
Comment (MrHDOLEK) [inline src/Services/NewsletterService.php:14]: please change magic number to  const  or another readable value

File: src/Notifications/NewsEmailNotification.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        return ["mail"];
+    }
+
+    public function toMail(): MailMessage
+    {
+        $title = $this->news->name;
+        $email = $this->subscriber->email;
+        $detailsUrl = route("meetups");
+
+        return (new MailMessage())
```
Comment (MrHDOLEK) [inline src/Notifications/NewsEmailNotification.php:33]: Maybe it is worth returning an already filled object?

File: src/Notifications/MeetupEmailNotification.php [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        return ["mail"];
+    }
+
+    public function toMail(): MailMessage
+    {
+        $title = $this->meetup->title;
+        $email = $this->subscriber->email;
+        $detailsUrl = route("meetups");
+
+        return (new MailMessage())
```
Comment (MrHDOLEK) [inline src/Notifications/MeetupEmailNotification.php:33]: Maybe it is worth returning an already filled object?

**Review by MrHDOLEK — CHANGES_REQUESTED** [review]

File: src/Models/News.php (hunk 10 above)
Comment (Majkulnn) [reply src/Models/News.php:11]: No, it's not, becouse it's News Model from #14  not Newsletter

File: src/Observers/NewsObserver.php [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+use Blumilk\Meetup\Core\Enums\AvailableNewsletter;
+use Blumilk\Meetup\Core\Models\News;
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+use Blumilk\Meetup\Core\Notifications\NewsEmailNotification;
+use Illuminate\Database\Eloquent\Collection;
+
+class NewsObserver
+{
+    public function __construct() {}
```
Comment (krzysztofrewak) [inline src/Observers/NewsObserver.php:15]: The same as above.

File: src/Services/NewsletterService.php [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Services;
+
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+use Illuminate\Http\Response;
+
+class NewsletterService
+{
+    public function isSubscriber(NewsletterSubscriber $subscriber): Response
```
Comment (krzysztofrewak) [inline src/Services/NewsletterService.php:12]: I don't like services returning HTTP responses. It should be reusable for every entrypoint.

File: src/Services/NewsletterService.php [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        if ($subscriber->getAttribute("status") === true) {
+            return response("You are already subscriber. Would you like to change your preferences?");
+        }
+        $subscriber->status = true;
```
Comment (krzysztofrewak) [inline src/Services/NewsletterService.php:17]: This doesn't mean anything. Maybe word `$status` could be replaced with something more verbose?

File: src/Http/Requests/Newsletter/NewsletterUpdateRequest.php [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Validation\Rule;
+
+class NewsletterUpdateRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            "email" => ["required", "string", "email", "max:255"],
+            "type" => ["required", "array", Rule::in([AvailableNewsletter::MEETUPS->value, AvailableNewsletter::NEWS->value])],
```
[the remark this answers is not in the document]
Comment (Majkulnn) [reply src/Http/Requests/Newsletter/NewsletterUpdateRequest.php]: i hope it's better now

File: src/Models/News.php (hunk 11 above)
Comment (kamilpiech97) [reply src/Models/News.php:35]: its relation to user model @MrHDOLEK

File: resources/views/newsletter/index.blade.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+            <div>
+                <label for="email">Email:</label>
+                <input type="email" name="email" id="email" value="{{ old('email') }}">
+                <x-input-error for="email"/>
+            </div>
+            <input type="submit" name="Subscribe" value="Subscribe" onclick="subscribe()">
+            <input type="submit" name="Unsubscribe" value="Unsubscribe" onclick="unsubscribe()">
+        </form>
+        <script>
+            form=document.getElementById("newsletterForm");
```
Comment (krzysztofrewak) [inline resources/views/newsletter/index.blade.php:22]:
```suggestion
            const form=document.getElementById("newsletterForm");
```

File: src/Enums/AvailableNewsletter.php [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+    case NEWS = "news";
+    case MEETUPS = "meetups";
+
+    public static function values(): array
+    {
+        $cases = collect(AvailableNewsletter::cases());
+
+        return $cases->map(
+            fn(AvailableNewsletter $enum) => $enum->value,
+        )->toArray();
```
Comment (krzysztofrewak) [inline src/Enums/AvailableNewsletter.php:18]:
```suggestion
        return array_map(self::cases(), fn(AvailableNewsletter $enum): string => $enum->value);
```

This should be quicker.

File: src/Services/NewsletterService.php [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Services;
+
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+
+class NewsletterService
+{
+    public function isSubscriber(NewsletterSubscriber $subscriber): string
```
Comment (krzysztofrewak) [inline src/Services/NewsletterService.php:11]: I really don't like idea that getter-by-name method is changing something and returning string message. Please refactor this to something more understandable.

File: src/Services/NewsletterService.php [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+        return "You are successfully subscribed";
+    }
+
+    public function unsubscribe(NewsletterSubscriber $subscriber): string
+    {
+        $subscriber->preferences()->delete();
+        $subscriber->subscription_state = false;
+        $subscriber->saveOrFail();
+
+        return "You have delete your subscription";
```
Comment (krzysztofrewak) [inline src/Services/NewsletterService.php:41]: Maybe it should be just void?

```suggestion
    public function unsubscribe(NewsletterSubscriber $subscriber): void
    {
        $subscriber->preferences()->delete();
        $subscriber->subscription_state = false;
        $subscriber->saveOrFail();
```

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: database/migrations/2022_03_31_162450_create_newsletter_preferences_table.php [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::create("newsletter_preferences", function (Blueprint $table): void {
+            $table->id();
+            $table->foreignId("newsletter_subscriber_id")
+                ->references("id")->on("newsletter_subscribers")
+                ->onDelete("cascade");
+            $table->string("preference");
```
Comment (mtracz) [inline database/migrations/2022_03_31_162450_create_newsletter_preferences_table.php:17]: What will be stored in this column?

File: src/Enums/AvailableNewsletter.php [hunk 24]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Blumilk\Meetup\Core\Enums;
+
+enum AvailableNewsletter: string
+{
+    case NEWS = "news";
+    case MEETUPS = "meetups";
```
Comment (mtracz) [inline src/Enums/AvailableNewsletter.php:10]: I think cases should be CamelCase. `News` and `Meetups`

As it is in PHP Docs:
https://www.php.net/manual/en/language.enumerations.backed.php

File: src/Http/Controllers/NewsletterSubscriberController.php [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+use Illuminate\View\View;
+
+class NewsletterSubscriberController extends Controller
+{
+    public function create(): View
+    {
+        return view("newsletter.index");
+    }
+
+    public function store(NewsletterStoreRequest $request, NewsletterService $service): View
```
Comment (mtracz) [inline src/Http/Controllers/NewsletterSubscriberController.php:20]: I think moving `NewsletterService` to the constructor will be better. You are using it 3 times.

File: src/Http/Controllers/NewsletterSubscriberController.php [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+    public function edit(): View
+    {
+        return view("newsletter.subscribe");
+    }
+
+    public function update(NewsletterUpdateRequest $request, NewsletterService $service, NewsletterSubscriber $subscriber): View
+    {
+        $subscriber = $subscriber->newQuery()->where("email", $request->validated("email"))->first();
+        $service->subscribe($subscriber);
+        $service->preference($subscriber, $request->validated("type"));
```
Comment (mtracz) [inline src/Http/Controllers/NewsletterSubscriberController.php:39]: What will happen if `$subscriber` will be `null`? `first()` method may return null.

Additionally, is `newQuery()` required here?

File: src/Enums/AvailableNewsletter.php [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+
+    public static function values(): array
+    {
+        return array_map(fn(AvailableNewsletter $enum): string => $enum->value, self::cases());
```
Comment (mtracz) [inline src/Enums/AvailableNewsletter.php:14]: Will be easier:
`array_column(self::cases(), 'value');`

File: src/Models/NewsletterPreferences.php [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+use Blumilk\Meetup\Core\Enums\AvailableNewsletter;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Support\Carbon;
+
+/**
+ * @property int $id
+ * @property string $preference
```
Comment (mtracz) [inline src/Models/NewsletterPreferences.php:15]: I think it should be enum instead of string.

File: src/Observers/MeetupObserver.php [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+use Blumilk\Meetup\Core\Notifications\MeetupEmailNotification;
+use Illuminate\Database\Eloquent\Collection;
+
+class MeetupObserver
+{
+    public function created(Meetup $meetup): void
+    {
+        $this->getSubscribersForNotifications()
+            ->each(fn($subscriber) => $subscriber->notify(new MeetupEmailNotification($meetup, $subscriber)));
```
Comment (mtracz) [inline src/Observers/MeetupObserver.php:18]: Please add types for arrow function.

File: src/Observers/NewsObserver.php [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+use Blumilk\Meetup\Core\Models\NewsletterSubscriber;
+use Blumilk\Meetup\Core\Notifications\NewsEmailNotification;
+use Illuminate\Database\Eloquent\Collection;
+
+class NewsObserver
+{
+    public function created(News $news): void
+    {
+        $this->getSubscribersForNotifications()
+            ->each(fn($subscriber) => $subscriber->notify(new NewsEmailNotification($news, $subscriber)));
```
Comment (mtracz) [inline src/Observers/NewsObserver.php:18]: Please add types for arrow function.

**Review by mtracz — CHANGES_REQUESTED** [review]

File: src/Services/NewsletterService.php [hunk 31]
```diff
@@ -8,10 +8,14 @@
...
+
     public function subscribe(NewsletterSubscriber $subscriber): void
     {
-        if (!$subscriber->subscription_state) {
-            $subscriber->subscription_state = true;
+        if (!$subscriber->subscribed) {
+            $subscriber->subscribed = true;
             $subscriber->saveOrFail();
         }
     }
```
Comment (mtracz) [inline src/Services/NewsletterService.php:21]: For what you need constructor? It is not in use.

File: src/Observers/NewsObserver.php [hunk 32]
```diff
@@ -15,13 +15,15 @@ class NewsObserver
     public function created(News $news): void
     {
         $this->getSubscribersForNotifications()
-            ->each(fn($subscriber) => $subscriber->notify(new NewsEmailNotification($news, $subscriber)));
+            ->each(function ($subscriber) use ($news): void {
+                $subscriber->notify(new NewsEmailNotification($news, $subscriber));
+            });
```
Comment (mtracz) [inline src/Observers/NewsObserver.php:20]: Arrow function was ok. Just missed types.
```php
fn(NewsletterSubscriber $subscriber): void => $subscriber->notify(new NewsEmailNotification($news, $subscriber))
```

File: src/Observers/MeetupObserver.php [hunk 33]
```diff
@@ -15,13 +15,15 @@ class MeetupObserver
     public function created(Meetup $meetup): void
     {
         $this->getSubscribersForNotifications()
-            ->each(fn($subscriber) => $subscriber->notify(new MeetupEmailNotification($meetup, $subscriber)));
+            ->each(function ($subscriber) use ($meetup): void {
+                $subscriber->notify(new MeetupEmailNotification($meetup, $subscriber));
+            });
```
Comment (mtracz) [inline src/Observers/MeetupObserver.php:20]: Back to arrow function and add types:
```php
fn(NewsletterSubscriber $subscriber): void => $subscriber->notify(new MeetupEmailNotification($meetup, $subscriber))
```

File: src/Models/NewsletterPreferences.php [hunk 34]
```diff
@@ -12,7 +12,7 @@
 
 /**
  * @property int $id
- * @property string $preference
+ * @property enum $preference
```
Comment (mtracz) [inline src/Models/NewsletterPreferences.php:15]: I mean use class name, in this case `AvailableNewsletter`

**Review by mtracz — CHANGES_REQUESTED** [review]

**Review by mtracz — APPROVED** [review]

**Review by krzysztofrewak — APPROVED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Review by MrHDOLEK — APPROVED** [review]

**Pull request merged**